netty/src/main/java/io/grpc/transport/netty/Http2Negotiator.java

```diff
@@ -104,26 +111,50 @@ public static ChannelHandler serverTls(SSLEngine sslEngine) {
-    Preconditions.checkNotNull(sslEngine, "sslEngine");
-    final SettableFuture<Void> completeFuture = SettableFuture.create();
-    if (!installJettyTlsProtocolSelection(sslEngine, completeFuture, false)) {
-      throw new IllegalStateException("NPN/ALPN extensions not installed");
-    }
-    SslHandler sslHandler = new SslHandler(sslEngine, false);
-    sslHandler.handshakeFuture().addListener(
-        new GenericFutureListener<Future<? super Channel>>() {
-          @Override
-          public void operationComplete(Future<? super Channel> future) throws Exception {
-            // If an error occurred during the handshake, throw it to the pipeline.
-            if (future.isSuccess()) {
-              completeFuture.get();
-            } else {
-              future.get();
-            }
+  public static ChannelHandler tls(final SslContext sslContext, final InetSocketAddress inetAddress,
+      final ChannelHandler handler) {
+    Preconditions.checkNotNull(sslContext, "sslContext");
+    Preconditions.checkNotNull(inetAddress, "inetAddress");
+
+    return new ChannelHandlerAdapter() {
+      @Override
+      public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+        SSLEngine sslEngine;
+        final SettableFuture<Void> completeFuture = SettableFuture.create();
+        if (isOpenSsl(sslContext.getClass())) {
+          sslEngine = sslContext.newEngine(ctx.alloc());
+          SSLParameters sslParams = new SSLParameters();
+          sslParams.setEndpointIdentificationAlgorithm("HTTPS");
+          sslEngine.setSSLParameters(sslParams);
+          completeFuture.set(null);
+        } else {
+          // Using JDK SSL
+          sslEngine
+              = sslContext.newEngine(ctx.alloc(), inetAddress.getHostName(), inetAddress.getPort());
+          SSLParameters sslParams = new SSLParameters();
+          sslParams.setEndpointIdentificationAlgorithm("HTTPS");
+          sslEngine.setSSLParameters(sslParams);
+          if (!installJettyTlsProtocolSelection(sslEngine, completeFuture, false)) {
+            throw new IllegalStateException("NPN/ALPN extensions not installed");
           }
-        });
-    return new BufferUntilTlsNegotiatedHandler(sslHandler, handler);
+        }
+        SslHandler sslHandler = new SslHandler(sslEngine, false);
+        sslHandler.handshakeFuture().addListener(
+            new GenericFutureListener<Future<? super Channel>>() {
+              @Override
+              public void operationComplete(Future<? super Channel> future) throws Exception {
+                // If an error occurred during the handshake, throw it to the pipeline.
+                if (future.isSuccess()) {
+                  completeFuture.get();
+                } else {
+                  future.get();
+                }
+              }
+            });
+        ChannelHandler tlsHandler = new BufferUntilTlsNegotiatedHandler(sslHandler, handler);
```

Comment:
It matters, a lot.

handlerAdded is called in the client's event loop, _not_ this one. When you call channel.write(), it finds the last ctx that matters within _this_ thread, and writes to it within the client's event loop. So if handlerAdded() hasn't been called yet and we call write(), then we end up writing to this handler, _even if it is replaced before the Runnable runs._ https://github.com/netty/netty/issues/3702 has the technical details of why.

It is even worse than that, though, because Netty detects that this handler doesn't care about writes. So instead of writing to _this_ handler, it writes to the _next_ handler which is the head of the pipeline.